tests/controllers/oauthClient.test.js

```diff
@@ -0,0 +1,103 @@
+const tape = require('tape')
+const supertest = require('supertest')
+const server = require('../../src/server.js')
+const Client = require('../../src/models/auth/Client.js')
+const User = require('../../src/models/User.js')
+
+const grants = require('../../src/constants/grants.js')
+
+const { addUserWithHashedPassword } = require('../helpers/index.js')
+const { makeLoggedInToken } = require('../../src/controllers/session.js')
+const { dropCollectionsAndEnd } = require('../helpers/index.js')
+const { validUser1 } = require('../fixtures/users.json')
+const { client, clientFromFormData } = require('../fixtures/auth/clients.json')
+
+tape('emptying db.', t => {
+  Promise.all([
+    User.remove({}),
+    Client.remove({})
+  ])
+  .then(() => t.end())
+  .catch(err => t.end(err))
+})
+
+tape('GET /oauth/clients when nothing in database', t => {
+  addUserWithHashedPassword(validUser1)
+  .then(() => makeLoggedInToken(validUser1))
+  .then(token => {
+    return supertest(server)
+    .get('/oauth/clients')
+    .set('Cookie', `token=${token}`)
+    .expect(200)
+  })
+  .then(res => {
+    t.equal(res.body.length, 0, 'should return length 1 array')
+    dropCollectionsAndEnd([Client, User], t)
+  })
+  .catch(err => t.end(err))
+})
+
+tape('filling db', t => {
+  Client.create(client)
+  .then(() => t.end())
+  .catch(err => t.end(err))
+})
+
+tape('GET /oauth/clients when something in database', t => {
+  addUserWithHashedPassword(validUser1)
+  .then(() => makeLoggedInToken(validUser1))
+  .then(token => {
+    return supertest(server)
+    .get('/oauth/clients')
+    .set('Cookie', `token=${token}`)
+    .expect(200)
+  })
+  .then(res => {
+    t.equal(res.body.length, 1, 'should return length 1 array')
+    dropCollectionsAndEnd([Client, User], t)
+  })
+  .catch(err => t.end(err))
+})
+
+tape('emptying db.', t => {
+  Promise.all([
+    User.remove({}),
+    Client.remove({})
+  ])
+  .then(() => t.end())
+  .catch(err => t.end(err))
+})
+
+tape('POST /oauth/clients', t => {
+  addUserWithHashedPassword(validUser1)
+  .then((user) => {
+    makeLoggedInToken(validUser1)
+    .then(token => {
+      return supertest(server)
+      .post('/oauth/clients')
+      .set('Cookie', `token=${token}`)
+      .send(clientFromFormData)
+      .expect(201)
+    })
+    .then(res => {
+      t.equal(res.body.name, clientFromFormData.name, 'correct name added')
+      t.equal(res.body.user, user.id, 'correct user added')
+      t.equal(res.body.secret.length, 16, '16 digit secret added')
+      t.deepEqual(res.body.redirectUris, clientFromFormData.redirectUris, 'correct redirectUris added')
+      t.deepEqual(res.body.grants, [grants.authCode], 'correct grants added')
+      t.ok(res.body._id, 'client given an id')
```

Comment:
Also look inside the database?

Reply:
I've done a simple check (findById, check the name is correct) to look inside the database for 2 reasons:

- 1. the first set of tests are just checking what's returned from a mongoose add function effectively, we should be able to be fairly certain that if something is returned and it matches the correct things then it is in the database.
- 2. When I tried to be more thorough I got stuff like this:
```
      operator: deepEqual
      expected: [ 'www.foundersandcoders.com' ]
      actual:   [ 'www.foundersandcoders.com' ]
      at: Client.findById.then.newClient (/Users/mdavis/Desktop/open-tourism-platform/tests/controllers/oauthClient.test.js:114:15)
      stack: |
```
Which is obviously tape just being weird, but as I don't see this as a massively important test (as it's just checking mongoose had worked properly) so I'm not gonna spend ages working out why right now. Also there's a few string conversions and stuff that I really don't think is necessary.

Reply:
Or maybe it's me misunderstanding deepEqual